### PR TITLE
tracer: add OpenTelemetry tracing with OPENTELEMETRY_DISABLED=false

### LINKED ISSUE #40589: trace/zoekt: export OpenTelemetry from Zoekt

### PR DESCRIPTION
**Pull request by bobheadxi**
Adds the ability for Zoekt to export OpenTelemetry, configured with `OPENTELEMETRY_DISABLED=false` (analogous to `JAEGER_DISABLED=false`). It updates `internal/trace` with mostly the same OpenTelemetry setup as `internal/trace` in `sourcegraph/sourcegraph` - it turns out most of the same workarounds are required for this to work correctly out of the box.

Closes https://github.com/sourcegraph/sourcegraph/issues/40589

Instructions for enabling in dev are currently in https://github.com/sourcegraph/sourcegraph/pull/40629, and I tested this by following the instructions in https://docs.sourcegraph.com/dev/how-to/zoekt_local_dev and running a search with tracing enabled. Inspecting the trace in Honeycomb confirms it went through the OpenTelemetry collector:

Screenshot 2022-08-19 at 2 49 50 PM

### CODE REVIEW

File: internal/otlpenv/otlpenv.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Package otlpenv exports getters to read OpenTelemetry protocol configuration options
+// based on the official spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options
+package otlpenv
+
+import (
+	"os"
+	"strings"
+)
+
+// getWithDefault returns def if no env in keys is set, or the first env from keys that is
```
Comment (jhchabran) [inline internal/otlpenv/otlpenv.go]:
```suggestion
// getWithDefault returns the default value if no env in keys is set, or the first env from keys that is
```

File: internal/otlpenv/otlpenv.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		if v, set := os.LookupEnv(k); set {
+			return v
+		}
+	}
+	return def
+}
+
+// This is a custom default that's also not quite compliant but hopefully close enough (we
+// use 127.0.0.1 instead of localhost, since there's a linter rule banning localhost).
+const defaultGRPCCollectorEndpoint = "http://127.0.0.1:4317"
```
Comment (jhchabran) [inline internal/otlpenv/otlpenv.go:23]: FWIW we can add exceptions to that linter, https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/check/no-localhost-guard.sh?L15-21

File: internal/tracer/jaeger.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		return nil, err
+	}
+	cfg.Tags = append(cfg.Tags, opentracing.Tag{Key: "service.version", Value: version})
+	if reflect.DeepEqual(cfg.Sampler, &jaegercfg.SamplerConfig{}) {
+		// Default sampler configuration for when it is not specified via
+		// JAEGER_SAMPLER_* env vars. In most cases, this is sufficient
+		// enough to connect to Jaeger without any env vars.
+		cfg.Sampler.Type = jaeger.SamplerTypeConst
+		cfg.Sampler.Param = 1
```
Comment (jhchabran) [inline internal/tracer/jaeger.go]: Probably worth adding a small comment on why it's `1` here.

**Review by jhchabran — APPROVED** [review]
🎉 (very minor fix, pre-approving, not worth going back and forth)

**Review by vrto — COMMENTED** [review]
Should we try creating a small lib so that Zoekt and Sourcegraph repos can share most of the OTel setting logic, or do they differ too much, in your opinion?

**Comment by bobheadxi** [comment]
> Should we try creating a small lib so that Zoekt and Sourcegraph repos can share most of the OTel setting logic, or do they differ too much, in your opinion?

I considered this, but the precedence here appears to be to do the copy-paste (also see the trace package, which is similarly copy-pasted). There are also variances like the configuration method (env variables vs site config) and dependency restrictions like the lack of `sourcegraph/log` (pending #411). tl;dr I think there are enough subtle differences to keep this copy-paste-ish for now, and consider reconciling them in the future - cc @keegancsmith

File: internal/otlpenv/otlpenv.go (hunk 2 above)
Comment (bobheadxi) [reply internal/otlpenv/otlpenv.go:23]: Yep! But there _is_ a rationale behind that, so I think it best to just leave this as is for now

File: internal/tracer/jaeger.go (hunk 3 above)
Comment (bobheadxi) [reply internal/tracer/jaeger.go]: I'm not sure either, this is just moved code (the jaeger setup here is unchanged) - I'll take a look and try to comment on it

Comment (bobheadxi) [reply internal/tracer/jaeger.go]: https://github.com/sourcegraph/zoekt/pull/413/commits/f87e6a2611c2164c3702b8e60fa18ef182852855

**Review by vrto — APPROVED** [review]

**Review by keegancsmith — APPROVED** [review]

**Comment by keegancsmith** [comment]
> I think there are enough subtle differences to keep this copy-paste-ish for now, and consider reconciling them in the future

Yup I think once the dust settles a lot of our common infra may make sense to put in a library. This stuff usually rarely changes :) Who knows, people other than sourcegraph may be interested. IE I am thinking debugserver, logging and tracing all make sense.

Out of interest, in the future will the integration be tiny for otel? It does feel like we have a surprising amount of code configuring our tracers these days, and hopefully if we only support otel the integration code should be tiny.

**Comment by keegancsmith** [comment]
Because I took a while to review this, I went ahead and merged in main for you so the conflicts are gone.

**Comment by bobheadxi** [comment]
> Out of interest, in the future will the integration be tiny for otel? It does feel like we have a surprising amount of code configuring our tracers these days, and hopefully if we only support otel the integration code should be tiny.

We can extend it further with third-party OTEL libraries opportunistically (e.g. https://github.com/sourcegraph/sourcegraph/pull/39300), but they should be small touch points.

We do unfortunately have a lot of code configuring our tracers - firstly because the opentelemetry-go library does not offer a great out-of-the-box setup experience, and secondly to support some of our legacy stuff. This is where some unification could be valuable (and where the setup might come in for other services as well)

**Pull request merged**